### PR TITLE
Add MoE load balancing loss to distillation

### PR DESCRIPTION
**Pull request by JamesDeng42**
# Description

This PR introduces support for Mixture of Experts (MoE) load balancing loss during the distillation workflow.

Key Changes
   1. NNX Intermediate Extraction (maxtext_utils.py & qwen3.py): 
      - Replaced legacy Linen self.sow(...) calls with native nnx.Intermediate(load_balance_loss) inside
   2. Distillation Strategy Updates (distillation_utils.py & train_distill.py):
      - Upgraded DistillationForwardOutput to carry the collected moe_lb_loss.
      - Updated CombinedDistillationStrategy to actively add the moe_lb_loss to the total_loss so the optimizer
        minimizes it.
      - Surfaced "distill/moe_lb_loss" to the metrics dictionary for TensorBoard logging and visibility.
   3. Model Mutability (models.py):
      - Automatically appended "intermediates" to the mutable_collections list during the Transformer's forward pass
        whenever load_balance_loss_weight > 0.0 to ensure NNX variables can successfully write to the state.

# Tests
Added "distill/moe_lb_loss" to the expected metrics keys in the test suite to prevent regressions in train_distill_test.py.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/models/models.py [hunk 1]
```diff
@@ -505,6 +505,8 @@ def __call__(
       mutable_collections.append("intermediates")
     if self.config.distill_beta > 0.0 and "intermediates" not in mutable_collections:
       mutable_collections.append("intermediates")
+    if self.config.load_balance_loss_weight > 0.0 and "intermediates" not in mutable_collections:
```
Comment (vlad-karp) [inline src/maxtext/models/models.py:511]: it is better to put some MoE naming here for clarity. 
Since this can be applied for MoE models only, can we move it to some MoE-only code?

File: src/maxtext/utils/maxtext_utils.py [hunk 2]
```diff
@@ -1081,6 +1081,29 @@ def get_intermediate_value(model, nested_key, default=None, clear=False):
         intermediate_value = model.decoder.layers["self_attention"][nested_key].get_value()[-1]
         if clear:
           del model.decoder.layers["self_attention"][nested_key]
+    case "moe_lb_loss":
+      # pylint: disable=import-outside-toplevel
+      from flax import nnx
```
Comment (ajkv-google) [inline src/maxtext/utils/maxtext_utils.py]: You could move this import statement to the top of the file to keep things consistent

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 3]
```diff
@@ -373,13 +375,19 @@ def compute_loss(
 
     total_loss = base_logit_loss + feature_loss
 
+    moe_lb_loss = jnp.array(0.0)
+    if student_output.moe_lb_loss is not None:
+      moe_lb_loss = student_output.moe_lb_loss
+      total_loss += moe_lb_loss
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py]: if we add to the total distillation loss, it is probably betetr to add it with some configurable weight factor, like we do for the other loss components

File: src/maxtext/utils/maxtext_utils.py [hunk 4]
```diff
@@ -1081,6 +1081,29 @@ def get_intermediate_value(model, nested_key, default=None, clear=False):
         intermediate_value = model.decoder.layers["self_attention"][nested_key].get_value()[-1]
         if clear:
           del model.decoder.layers["self_attention"][nested_key]
+    case "moe_lb_loss":
```
Comment (gagika) [inline src/maxtext/utils/maxtext_utils.py]: Could you reuse 
                           
   moe_lb_loss = maxtext_utils.get_intermediate_value(model, "moe_lb_loss", clear=False)                                                         
  or                                                                           
  moe_lb_losses = maxtext_utils.collect_intermediates_by_suffix(intermediate_outputs, "moe_lb_loss")

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 3 above)
Comment (gagika) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: we should reuse `config.load_balance_loss_weight` for that, I see it's used in if statements.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 5]
```diff
@@ -373,13 +375,19 @@ def compute_loss(
 
     total_loss = base_logit_loss + feature_loss
 
+    moe_lb_loss = jnp.array(0.0)
+    if student_output.moe_lb_loss is not None:
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py:486]: seams like you are computing `moe_lb_loss` for both student and teacher, only student one is used for loss (as you are doing in this PR), but as you are computing it for teacher, we can also log it in tensorboard (distill/teacher_moe_lb_loss) so it can help debugging (comparing student's loss with teacher's).

File: src/maxtext/utils/maxtext_utils.py (hunk 2 above)
Comment (JamesDeng42) [reply src/maxtext/utils/maxtext_utils.py]: thanks for the catch.

File: src/maxtext/models/models.py (hunk 1 above)
Comment (JamesDeng42) [reply src/maxtext/models/models.py:511]: the load_balance_loss_weight is used other modules of maxtext as well and i just add this existed parameter to distillation pipeline, this is good suggestion but we probably should not change the name in this pr.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 3 above)
Comment (JamesDeng42) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: sorry, i forget this. let's add the load_balance_loss_weight before  add to the total distillation loss. thanks for the catch.

File: src/maxtext/utils/maxtext_utils.py (hunk 4 above)
Comment (JamesDeng42) [reply src/maxtext/utils/maxtext_utils.py]: thanks, it help remove this large ugly code block.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 5 above)
Comment (JamesDeng42) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py:486]: added, thanks for the suggestion.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 3 above)
Comment (vlad-karp) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: @JamesDeng42 , looks like you forgot to use load_balance_loss_weight and still just add the balancer loss to total loss as before

**Review by vlad-karp — COMMENTED** [review]
please address the comment

File: src/maxtext/utils/maxtext_utils.py (hunk 4 above)
Comment (gagika) [reply src/maxtext/utils/maxtext_utils.py]: can this whole case go away?

by calling collect_intermediates_by_suffix directly from train_distill.py instead, matching
how pre_train/train.py:249 and dpo_utils.py:139 already collect MoE LB loss:

```
intermediate_outputs = nnx.state(model, nnx.Intermediate)                                                                 
moe_lb_losses = maxtext_utils.collect_intermediates_by_suffix(intermediate_outputs, "moe_lb_loss")
if moe_lb_losses:                                                                                                         
      moe_lb_loss = jnp.mean(jnp.concatenate(moe_lb_losses))
```

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 3 above)
Comment (JamesDeng42) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: forget to push this change, added "total_loss += moe_lb_loss * self.load_balance_loss_weight"

File: src/maxtext/utils/maxtext_utils.py (hunk 4 above)
Comment (JamesDeng42) [reply src/maxtext/utils/maxtext_utils.py]: nice, yes.

Comment (JamesDeng42) [reply src/maxtext/utils/maxtext_utils.py]: yes, thanks a lot.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 6]
```diff
@@ -480,13 +484,24 @@ def compute_loss(
 
     total_loss = base_logit_loss + feature_loss
 
+    moe_lb_loss = jnp.array(0.0)
+    if student_output.moe_lb_loss is not None and self.load_balance_loss_weight > 0.0:
+      moe_lb_loss = student_output.moe_lb_loss
+      total_loss += moe_lb_loss * self.load_balance_loss_weight
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py]: sorry, for double comment, in moe.py, seams load_balance_loss_weight is already applied,

```
  # See Switch Transformer (https://arxiv.org/abs/2101.03961) for more details.
  def load_balance_loss(self, top_k_indices, logits) -> jax.Array:
    """Compute the load balance loss."""
    expert_mask = jax.nn.one_hot(top_k_indices, num_classes=self.num_experts, dtype=jnp.int32)
    summed_expert_mask = jnp.sum(expert_mask, axis=2)
    # Get fraction of tokens dispatched to each expert
    density = jnp.mean(summed_expert_mask, axis=1)
    # get fraction of probability allocated to each expert
    density_prob = jnp.mean(logits, axis=1)
    loss = jnp.mean(density * density_prob) * (self.num_experts**2) * self.config.load_balance_loss_weight
    return loss
```

could you double check to make sure we don't apply it twice? 

e.g. you can print `student_output.moe_lb_loss` and decrease self.load_balance_loss_weight see if moe_lb_loss decreases proportionally?

Comment (JamesDeng42) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: yes, wrote an unit test myself and it is applied twice, remove the extra.

Comment (gagika) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: thanks, could you comment in the code, so it's clear.

File: src/maxtext/models/qwen3.py [hunk 7]
```diff
@@ -1055,7 +1055,7 @@ def __call__(
     # We sow the load balancing loss so it can be collected and added to the total loss
     # during training.
     if self.config.load_balance_loss_weight > 0.0 and load_balance_loss is not None:
-      self.sow("intermediates", "moe_lb_loss", load_balance_loss)
+      self.moe_lb_loss = nnx.Intermediate(load_balance_loss)
```
Comment (gagika) [inline src/maxtext/models/qwen3.py:1058]: will other models also need to migrate to `nnx.Intermediate` to be used in distillation router load balancing?

File: src/maxtext/trainers/post_train/distillation/train_distill.py [hunk 8]
```diff
@@ -150,8 +150,17 @@ def model_forward_fn(
     if config.distill_beta > 0.0:
       out_projection_activations = maxtext_utils.get_intermediate_value(model, "out_projection_activations", clear=True)
 
+    moe_lb_loss = None
+    if config.num_experts > 1 and config.load_balance_loss_weight > 0.0:
+      intermediate_outputs = nnx.state(model, nnx.Intermediate)
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/train_distill.py:155]: nit: you could change this to pop and remove pop from below:

intermediates = nnx.pop(model, nnx.Intermediate)

File: src/maxtext/models/qwen3.py (hunk 7 above)
Comment (JamesDeng42) [reply src/maxtext/models/qwen3.py:1058]: No, most of the older models in MaxText are still technically using Flax Linen components.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 6 above)
Comment (JamesDeng42) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: added, thanks.

File: src/maxtext/trainers/post_train/distillation/train_distill.py (hunk 8 above)
Comment (JamesDeng42) [reply src/maxtext/trainers/post_train/distillation/train_distill.py]: thanks for the suggestions, applied. 👍

**Review by gagika — APPROVED** [review]
thanks

**Review by ajkv-google — APPROVED** [review]
